python/ray/tune/ray_trial.py

```diff
@@ -0,0 +1,288 @@
+        except Exception:
+            print("Error stopping runner:", traceback.format_exc())
+            self.status = Trial.ERROR
+        finally:
+            self.runner = None
+
+        if stop_logger and self.result_logger:
+            self.result_logger.close()
+            self.result_logger = None
+
+    def train_remote(self):
+        """Returns Ray future for one iteration of training."""
+
+        assert self.status == Trial.RUNNING, self.status
+        return self.runner.train.remote()
+
+    def save_to_object(self):
+        return self.runner.save_to_object.remote()
+
+    def save_to_path(self):
+        ray.get(self.runner.save.remote())
+
+    def _setup_runner(self):
+        self.status = Trial.RUNNING
+        cls = ray.remote(
+            num_cpus=self.resources.cpu,
+            num_gpus=self.resources.gpu)(self._get_trainable_cls())
+        if not self.result_logger:
+            if not os.path.exists(self.local_dir):
+                os.makedirs(self.local_dir)
+            self.logdir = tempfile.mkdtemp(
+                prefix="{}_{}".format(
+                    str(self)[:MAX_LEN_IDENTIFIER], date_str()),
+                dir=self.local_dir)
+            self.result_logger = UnifiedLogger(self.config, self.logdir,
+                                               self.upload_dir)
+        remote_logdir = self.logdir
+
+        def logger_creator(config):
+            # Set the working dir in the remote process, for user file writes
+            if not os.path.exists(remote_logdir):
+                os.makedirs(remote_logdir)
+            os.chdir(remote_logdir)
+            return NoopLogger(config, remote_logdir)
+
+        # Logging for trials is handled centrally by TrialRunner, so
+        # configure the remote runner to use a noop-logger.
+        self.runner = cls.remote(
+            config=self.config, logger_creator=logger_creator)
+
+    def restore_from_path(self, path):
+        """Restores runner state from specified path.
+
+        Args:
+            path (str): A path where state will be restored.
+        """
+
+        if self.runner is None:
+            print("Unable to restore - no runner")
+        else:
+            try:
+                ray.get(self.runner.restore.remote(path))
+            except Exception:
+                print("Error restoring runner:", traceback.format_exc())
+                self.status = Trial.ERROR
+
+    def restore_from_obj(self, obj):
+        """Restores runner state from the specified object."""
+
+        if self.runner is None:
+            print("Unable to restore - no runner")
+        else:
+            try:
+                ray.get(self.runner.restore_from_object.remote(obj))
+            except Exception:
+                print("Error restoring runner:", traceback.format_exc())
+                self.status = Trial.ERROR
+
+register_trial_impl(RayTrial) # any call to Trial() will create RayTrial instead of Trial.
```

Comment:
this seems quite fragile... Ideally, this switch will be encapsulated in RayTrialExecutor or TrialRunner